src/Language/Printer.php

```diff
@@ -3,54 +3,373 @@
+     * @return string
+     */
+    protected function printVariable(VariableNode $node): string
+    {
+        return '$' . $node->getName();
     }
 
+    // Document
+
     /**
-     * @inheritdoc
+     * @param DocumentNode $node
+     * @return string
+     */
+    protected function printDocument(DocumentNode $node): string
+    {
+        return \implode("\n\n", $node->getDefinitions()) . "\n";
+    }
+
+    /**
+     * @param OperationDefinitionNode $node
+     * @return string
      * @throws LanguageException
      */
-    public function print(NodeInterface $node): string
+    protected function printOperationDefinition(OperationDefinitionNode $node): string
     {
-        if (($writer = $this->getWriter($node)) !== null) {
-            return $writer->write($node);
-        }
+        $operation            = $node->getOperation();
+        $name                 = $this->one($node->getName());
+        $variablesDefinitions = $this->many($node->getVariableDefinitions());
+        $directives           = $this->many($node->getDirectives());
+        $selectionSet         = $this->one($node->getSelectionSet());
 
-        throw new LanguageException(sprintf('Invalid AST Node: %s', (string)$node));
+        // Anonymous queries with no directives or variable definitions can use
+        // the query short form.
+        return null === $name && empty($directives) && empty($variablesDefinitions) && $operation === 'query'
+            ? $selectionSet
+            : \implode(' ', [
+                $operation,
+                $name . wrap('(', \implode(', ', $variablesDefinitions), ')'),
+                \implode(' ', $directives),
+                $selectionSet,
+            ]);
     }
 
     /**
-     * @param NodeInterface $node
-     * @return WriterInterface|null
+     * @param VariableDefinitionNode $node
+     * @return string
+     * @throws LanguageException
      */
-    protected function getWriter(NodeInterface $node): ?WriterInterface
+    protected function printVariableDefinition(VariableDefinitionNode $node): string
     {
-        foreach ($this->writers as $writer) {
-            if ($writer instanceof WriterInterface && $writer->supportsWriter($node)) {
-                return $writer;
-            }
-        }
+        $variable     = $this->one($node->getVariable());
+        $type         = $this->one($node->getType());
+        $defaultValue = $this->one($node->getDefaultValue());
+
+        return $variable . ': ' . $type . wrap(' = ', $defaultValue);
+    }
+
+    /**
+     * @param SelectionSetNode $node
+     * @return string
+     */
+    protected function printSelectionSet(SelectionSetNode $node): string
+    {
+        return block($this->many($node->getSelections()));
+    }
+
+    /**
+     * @param FieldNode $node
+     * @return string
+     * @throws LanguageException
+     */
+    protected function printField(FieldNode $node): string
+    {
+        $alias        = $this->one($node->getAlias());
```

Comment:
When I read `$this->one`, `$this->many` I don't really know what they do until I see the functions. I think `printNode`, `printNodes` are maybe easier to understand.